### PR TITLE
Split off all individual OverviewScreen panes

### PR DESCRIPTION
**Pull request by SomeTroglodyte**
Observations on mobile portrait mode (desktop too when resized accordingly):
- Large vertical portion of screen unused
- Cannot horizontally scroll far enough to see whole unit / resources pages
- Cannot horizontally scroll categories
- Flipping orientation while open...

Couldn't resist breaking it up.

- Every single category in its own file
- Consistent interface in case we might want to use a common superclass in some future (reason DiplomacyOverviewTable changed its parameter signature and gets one height indirectly now)

Other than that and the band-aids mentioned I just moved code around and adapted it (mostly table.add -> add in init{} blocks, but also a stage ref suddenly meaning a different stage needed a qualifier).

The band-aids:
1: Large vertical portion of screen unused -> height calculation centralized and changed from stage*0.8 to stage - room used outside
2: Cannot horizontally scroll far enough -> set width explicitly on the Cell containing the ScrollPane
3: Cannot horizontally scroll categories -> Added one ScrollPane
4: Added resize event override

To-do:
- One todo comment: Companion.getCivGroup seems quite mis-placed
- Units would look better if the header did not scroll off the top - nested ScrollPanes shouldn't be hard.
- Vertical variant of resources to satisfy e.g. still-open #2619
- Diplomacy status scaling taking narrow width into account

### CODE REVIEW

**Comment by yairm210** [comment]
Pics please?

**Comment by yairm210** [comment]
Since it's hard to see where changes were made, I'm going to assume everything's fine as long as it visually seems OK

File: core/src/com/unciv/ui/overviewscreen/EmpireOverviewScreen.kt [hunk 1]
```diff
@@ -99,391 +94,25 @@ class EmpireOverviewScreen(private var viewingPlayer:CivilizationInfo, defaultPa
         topTable.add(closeButton)
 
         topTable.pack()
+        val topScroll = ScrollPane(topTable).apply { setScrollingDisabled(false, true) }
```
Comment (SomeTroglodyte) [inline core/src/com/unciv/ui/overviewscreen/EmpireOverviewScreen.kt:97]: Band aid 1 (with 5 lines down)

File: core/src/com/unciv/ui/overviewscreen/EmpireOverviewScreen.kt [hunk 2]
```diff
@@ -1,37 +1,29 @@
...
     private val topTable = Table().apply { defaults().pad(10f) }
     private val centerTable = Table().apply { defaults().pad(5f) }
 
-    private val setCategoryActions = HashMap<String, () -> Unit>()
+    internal val setCategoryActions = HashMap<String, () -> Unit>()
     private val categoryButtons = HashMap<String, Button>()
 
+    // 50 normal button height + 2*10 topTable padding + 2 Separator + 2*5 centerTable padding
+    // Since a resize recreates this screen this should be fine as a val
+    internal val centerAreaHeight = stage.height - 82f
```
Comment (SomeTroglodyte) [inline core/src/com/unciv/ui/overviewscreen/EmpireOverviewScreen.kt:25]: Band Aid 3 (with all uses)

File: core/src/com/unciv/ui/overviewscreen/EmpireOverviewScreen.kt [hunk 3]
```diff
@@ -51,10 +43,13 @@ class EmpireOverviewScreen(private var viewingPlayer:CivilizationInfo, defaultPa
         val setCategoryAction = {
             centerTable.clear()
             centerTable.add(ScrollPane(table).apply { setOverscroll(false, false) })
-                    .height(stage.height * 0.8f)
+                    .height(centerAreaHeight)
+                    .width(stage.width)
```
Comment (SomeTroglodyte) [inline core/src/com/unciv/ui/overviewscreen/EmpireOverviewScreen.kt:47]: Band Aid 2

File: core/src/com/unciv/ui/overviewscreen/EmpireOverviewScreen.kt [hunk 4]
```diff
@@ -99,391 +94,25 @@ class EmpireOverviewScreen(private var viewingPlayer:CivilizationInfo, defaultPa
...
-            if (unit.canUpgrade()) promotionsTable.add(ImageGetter.getUnitIcon(unit.getUnitToUpgradeTo().name, Color.GREEN)).size(28f).padLeft(8f)
-            promotionsTable.onClick {
-                if (unit.promotions.canBePromoted() || unit.promotions.promotions.isNotEmpty()) {
-                    game.setScreen(PromotionPickerScreen(unit))
-                }
-            }
-            table.add(promotionsTable)
-            if (unit.health < 100) table.add(unit.health.toString()) else table.add()
-            table.row()
+    override fun resize(width: Int, height: Int) {
```
Comment (SomeTroglodyte) [inline core/src/com/unciv/ui/overviewscreen/EmpireOverviewScreen.kt:109]: Band Aid 4

File: core/src/com/unciv/ui/overviewscreen/EmpireOverviewScreen.kt [hunk 5]
```diff
@@ -51,10 +43,13 @@ class EmpireOverviewScreen(private var viewingPlayer:CivilizationInfo, defaultPa
         val setCategoryAction = {
             centerTable.clear()
             centerTable.add(ScrollPane(table).apply { setOverscroll(false, false) })
-                    .height(stage.height * 0.8f)
+                    .height(centerAreaHeight)
+                    .width(stage.width)
             centerTable.pack()
             for ((key, categoryButton) in categoryButtons.filterNot { it.value.touchable == Touchable.disabled })
                 categoryButton.color = if (key == name) Color.BLUE else Color.WHITE
+            if (name == "Stats")
```
Comment (SomeTroglodyte) [inline core/src/com/unciv/ui/overviewscreen/EmpireOverviewScreen.kt:51]: Band Aid 5

**Comment by SomeTroglodyte** [comment]
Sorry for not answering sooner. Outdoors called.

Recap: The _bulk_ of this PR is separating the pages/panes/tabs of the Overview into separate source files, like you did for the City page a year ago, for maintainability and readability.

Based on this, I thought no screenshots necessary, but you're right, I mixed in some provisional band-aids, and those can be shown (and I marked the code lines)-
"Band Aid"s 1 and 2:
image
Band Aid 3:
image

Band Aid 4 is the `override fun resize` - you know that one, no screenie

Band Aid 5 I forgot in the original PR comment - too lazy to screenshot that - the tutorial "look at your stats" was previously marked as fulfilled when the corresponding Table was built (which is well before display) - I moved that to where the category is actually displayed.

The rest of the diff is my attempt to faithfully change _no_ behaviour while still spearating the categories and refactoring them into a subclass instead of a factory method.

**Pull request merged**